### PR TITLE
[FSDP2] Added gradient accumulation w/o reduction

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119551
* #119378
* #119302
* __->__ #118298
* #119825
* #118755
* #118223
* #118136
* #119550

This PR adds a way to do gradient accumulation without collectives (i.e. reduce-scatter for FSDP and reduce-scatter/all-reduce for HSDP, though HSDP is not yet implemented). Since the `no_sync()` context manager has received some feedback, we simply define a method on the module to set whether the module requires gradient synchronization or not, where this method can recurse or not.
```
# Before with `no_sync()`:
with fsdp_model.no_sync() if not is_last_microbatch else contextlib.nullcontext():
  # Forward/backward

# After with a setter:
fsdp_model.set_requires_gradient_sync(not is_last_microbatch)
# Forward/backward
```
Having the method be able to recurse or not also gives some flexibility. For example, some large modules can still reduce-scatter, while some smaller modules can avoid it to save communication bandwidth:
```
fsdp_modules_to_reduce_scatter: Set[nn.Module] = ...
for module in fsdp_model.modules():
  if isinstance(module, FSDP) and module not in fsdp_modules_to_reduce_scatter:
    module.set_requires_gradient_sync(not is_last_microbatch)
# Forward/backward
```

(Separately, we may expose a helper for `return [module for model.modules() if isinstance(module, FSDP)]`.)

---

To show the spirit of this API choice, I also included `set_requires_all_reduce` that would give us the ability to only reduce-scatter but not all-reduce for HSDP (originally from the MiCS paper). If we want to flexibly support heterogeneous sharding where FSDP is applied to some modules and HSDP to others in the same model, then having a module-level method that has the option to not recurse makes sense to me.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Review by wconstab — APPROVED** [review]
this lgtm. seems sufficient for the PP cases I am familiar with.

i think we should introduce a consistent API for replicate / DDP.

File: torch/distributed/_composable/fsdp/fully_shard.py [hunk 1]
```diff
@@ -163,6 +163,39 @@ def set_is_last_backward(self, is_last_backward: bool) -> None:
         state = self._get_fsdp_state()
         state._state_ctx.is_last_backward = is_last_backward
 
+    def set_requires_gradient_sync(
+        self, requires_gradient_sync: bool, recurse: bool = True
+    ) -> None:
+        """
+        Sets if the module should sync gradients. This can be used to implement
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/fully_shard.py:170]: hmmm I wonder if we should allow user to totally turn off grad sync without turn the grad sync back on later?

FSDP normally only have two modes: 1. fully shard 2. HSDP, for either case, we need to do reduce_scatter and all_reduce. But if user set the `requires_gradient_sync=False` and forget to reduce the grads, then FSDP won't do anything for the model grads, wouldn't that make DP semantic be wrong?

Making user control whether to sync grad or not maybe necessary to compose with PP but I wonder if we should make this sth like a context manager and restore reduce_grad somehow later

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 2]
```diff
@@ -275,6 +280,9 @@ def pre_backward(self, forward_grad_fns: Tuple[Any, ...], *unused: Any):
     def post_backward(self, *unused: Any):
         self._training_state = TrainingState.POST_BACKWARD
         with torch.profiler.record_function("FSDP::post_backward_reshard"):
+            if not self.reduce_scatter_grads:
```
Comment (wanchaol) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:283]: `if not self.reduce_scatter_grads and self.all_reduce_grads` we should also skip reshard?

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/fully_shard.py:170]: My reasoning for a setter instead of a context is as follows:
1. Whether a setter or a context, the usage is conventionally a function of `is_last_microbatch`, which is computed from the microbatch index. There should not be a risk of forgetting to re-enable this gradient sync.
    - The expected usage of the setter is `module.set_requires_gradient_sync(is_last_microbatch)`, not directly as `set_requires_gradient_sync(True)` or `set_requires_gradient_sync(False)`.
    - The current usage of `no_sync()` looks like `with root_module.no_sync() if not is_last_microbatch else contextlib.nullcontext()`.
2. Using a context manager makes it more clumsy to selectively disable gradient sync for some submodules but not others. The existing `no_sync()` must be called on the root FSDP module and applies to all child FSDP modules as well. However, we can imagine performance tuning where, when accumulating gradients across microbatches, we disable gradient sync for some submodules but not others to trade-off communication and memory.
3. `no_sync()` is not always intuitive. There have been issues where DDP users did not know that they must include both `forward()` and `backward()` under the context, leading to incorrect behavior (https://github.com/pytorch/pytorch/pull/89244). In contrast, FSDP does not need both `forward()` and `backward()` to be under the context; however, users may extrapolate from DDP. Karpathy gave this feedback in nanoGPT, where instead prefers setting a variable:
https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/train.py#L291-L295

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 2 above)
Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:283]: We disallow the case of no reduce-scatter but all-reduce. For HSDP, we either do not gradient sync (no reduce-scatter, no all-reduce) or just reduce-scatter (reduce-scatter, no all-reduce). If `not self.reduce_scatter_grads`, then that implies `not self.all_reduce_grads`.

File: torch/distributed/_composable/fsdp/fully_shard.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/_composable/fsdp/fully_shard.py:170]: Ok yeah this convinced me!

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:283]: so this check would be set whenever we initialize the HSDP mode? (i.e. `self.reduce_scatter_grads==self.all_reduce_grads`)

Comment (awgu) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:283]: Yes. It would be an invariant that we preserve.

**Review by wanchaol — APPROVED** [review]
lgtm
